wpical/src/main/native/cpp/fieldcalibration.cpp

```diff
@@ -0,0 +1,824 @@
+      constraint.t_begin_end.q =
+          Eigen::Quaterniond(tag_min_to_tag.block<3, 3>(0, 0));
+
+      constraints.push_back(constraint);
+    }
+
+    apriltag_detections_destroy(tag_detections);
+
+    // Show debug
+    cv::imshow("Processing Frame", frame_debug);
+    cv::waitKey(1);
+  }
+
+  video_input.release();
+  cv::destroyAllWindows();
+
+  return true;
+}
+
+int fieldcalibration::calibrate(std::string input_dir_path, std::string output_file_path,
+                                std::string camera_model_path, std::string ideal_map_path,
+                                int pinned_tag_id, int detection_fps)
+{
+
+  // Silence OpenCV logging
+  cv::utils::logging::setLogLevel(
+      cv::utils::logging::LogLevel::LOG_LEVEL_SILENT);
+
+  // Load camera model
+  const auto [camera_matrix, camera_distortion] =
+      load_camera_model(camera_model_path);
+
+  // Load ideal field map
+  std::map<int, nlohmann::json> ideal_map = load_ideal_map(ideal_map_path);
+
+  // Apriltag detector
+  apriltag_detector_t *tag_detector = apriltag_detector_create();
+  tag_detector->nthreads = 8;
+
+  apriltag_family_t *tag_family = tag36h11_create();
+  apriltag_detector_add_family(tag_detector, tag_family);
+
+  // Find tag poses
+  std::map<int, Pose, std::less<int>,
+           Eigen::aligned_allocator<std::pair<const int, Pose>>>
+      poses;
+  std::vector<Constraint, Eigen::aligned_allocator<Constraint>> constraints;
+
+  for (const auto &entry :
+       std::filesystem::directory_iterator(input_dir_path))
+  {
+    if (entry.path().filename().string()[0] == '.')
+    {
+      continue;
+    }
+
+    const std::string path = entry.path().string();
+
+    bool success =
+        process_video_file(tag_detector, detection_fps, camera_matrix,
+                           camera_distortion, 0.1651, path, poses, constraints);
+
+    if (!success)
+    {
+      std::cout << "Unable to process video" << std::endl;
+      return -1;
+    }
+  }
+
+  // Build optimization problem
+  ceres::Problem problem;
+  ceres::Manifold *quaternion_manifold = new ceres::EigenQuaternionManifold;
+
+  for (const auto &constraint : constraints)
+  {
+    auto pose_begin_iter = poses.find(constraint.id_begin);
+    auto pose_end_iter = poses.find(constraint.id_end);
+
+    ceres::CostFunction *cost_function =
+        PoseGraphError::Create(constraint.t_begin_end);
+
+    problem.AddResidualBlock(cost_function, nullptr,
+                             pose_begin_iter->second.p.data(),
+                             pose_begin_iter->second.q.coeffs().data(),
+                             pose_end_iter->second.p.data(),
+                             pose_end_iter->second.q.coeffs().data());
+
+    problem.SetManifold(pose_begin_iter->second.q.coeffs().data(),
+                        quaternion_manifold);
+    problem.SetManifold(pose_end_iter->second.q.coeffs().data(),
+                        quaternion_manifold);
```

Comment:
Consider reforumulating this as a cost function on tag corner reprojection error rather than pose error. This allows you to side-step pose ambiguity introduced by solvePNP. See https://github.com/mcm001/gtsam-playground/blob/a2de73d2fdb2c4feb39ea068dcfb7d333685babc/src/gtsam_utils.cpp#L43 & https://github.com/mcm001/gtsam-playground/blob/a2de73d2fdb2c4feb39ea068dcfb7d333685babc/src/localizer.cpp#L412, but basically the formulation here is:

```
for tag in tags:
// for each tag corner (4 per tag)
for i in range(4)
// Vec3 from camera CS to tag corner point
cameraTtag_corner = worldTtag_corner_i.relativeTo(worldTcamera_i)
// Vec2 in pixels
predicted_corner_px = camera_model.project(cameraTtag_corner )
error = predicted_corner_px - measured_corner_px
```

Reply:
This could also be formulated as a factor graph using gtsam - this is a specialization of a generic structure from motion problem (https://github.com/borglab/gtsam/blob/4.2.0/examples/SFMExampleExpressions.cpp) but I think that in both cases with the same noise model used for all tag corner pixel measurements, the math will reduce to a least squares optimization problem on reprojection error and I'd expect you to get the same output from both.

Reply:
Other prior art: https://github.com/berndpfrommer/tagslam/blob/a8afd5d2bb85e8d783831af7cedc0838e0ed53d1/src/gtsam_optimizer.cpp#L205

Reply:
https://github.com/mcm001/gtsam-playground/blob/e273fc41560ee0f88897a01c3b429cbe564f1104/src/sfm_mapper/main.cpp#L119C1-L138C6 is a snippet of a working implementation of this in gtsam.